packages/flutter/lib/src/material/search.dart

```diff
@@ -0,0 +1,416 @@
+/// `delegate`. The initial query string is given by `query`, which defaults
+/// to the empty string. When `query` is set to null, `delegate.query` will
+/// be used as the initial query.
+///
+/// The method returns the selected search result, which can be set in the
+/// [SearchDelegate.close] call.
+///
+/// The transition to the search page triggered by this method looks best if the
+/// screen triggering the transition contains an [AppBar] at the top and the
+/// transition is called from an [IconButton] that's part of [AppBar.actions].
+/// The animation provided by [SearchDelegate.transitionAnimation] can be used
+/// to trigger additional animations in the underlying page while the search
+/// page fades in or out. This is commonly used to animate an [AnimatedIcon] in
+/// the [AppBar.leading] position e.g. from the hamburger menu to the back arrow
+/// used to exit the search page.
+///
+/// See also:
+///
+///  * [SearchDelegate] to define the content of the search page.
+Future<T> showSearch<T>({
+  @required BuildContext context,
+  @required SearchDelegate<T> delegate,
+  String query: '',
+}) {
+  assert(delegate != null);
+  assert(context != null);
+  assert(delegate._result == null || delegate._result.isCompleted);
+  delegate._result = new Completer<T>();
+  delegate.query = query ?? delegate.query;
+  delegate._currentBody.value = SearchBody.suggestions;
+  Navigator.of(context).push(new _SearchPageRoute<T>(
+    delegate: delegate,
+  ));
+  return delegate._result.future;
+}
+
+/// Delegate for [showSearch] to define the content of the search page.
+///
+/// The search page always shows an [AppBar] at the top where users can
+/// enter their search queries. The buttons shown before and after the search
+/// query text field can be customized via [SearchDelegate.leading] and
+/// [SearchDelegate.actions].
+///
+/// The body below the [AppBar] can either show suggested queries (returned by
+/// [SearchDelegate.buildSuggestions]) or - once the user submits a search  - the
+/// results of the search as returned by [SearchDelegate.buildResults].
+///
+/// [SearchDelegate.query] always contains the current query entered by the user
+/// and should be used to build the suggestions and results.
+///
+/// The results can be brought on screen by calling [SearchDelegate.showResults]
+/// and you can go back to showing the suggestions by calling
+/// [SearchDelegate.showSuggestions].
+///
+/// Once the user has selected a search result, [SearchDelegate.close] should be
+/// called to remove the search page from the top of the navigation stack and
+/// to notify the caller of [showSearch] about the selected search result.
+abstract class SearchDelegate<T> {
+
+  /// Suggestions shown in the body of the search page while the user types a
+  /// query into the search field.
+  ///
+  /// The delegate method is called whenever the content of [query] changes and
+  /// the value of [query] can be used to determine which suggestions should
```

Comment:
Since I think we really intend for the suggestions to be based on the query string: The suggestions should be based on the [query] string ... And I think you should what the method is supposed to do if [query] is the empty string, since that's the default.